implementations/elixir/ockam/ockam/lib/ockam/secure_channel/channel.ex

```diff
@@ -1,380 +1,616 @@
+      # that initiator need to finish the handshake by sending proving his own Identity  too.
+      case ie.waiting do
+        {pid, ref} ->
+          send(pid, {:connected, ref})
+
+        nil ->
+          :ok
+      end
+
+      {:ok, :cancel} = :timer.cancel(ie.timer)
+
+      established = %Established{
+        encrypt_st: ie.encrypt_st,
+        decrypt_st: ie.decrypt_st,
+        h: ie.h,
+        peer_identity: identity,
+        peer_identity_id: identity_id
+      }
 
-      :error ->
-        case Ockam.Vault.Software.init() do
-          {:ok, vault} ->
-            vault
+      case state.role do
+        :initiator ->
+          state = %Channel{state | peer_identity_route: msg.return_route}
 
-          _error ->
-            nil
-        end
+          {:ok, encrypt_st} =
+            send_identity_proof(
+              IdentityHandshake.Response,
+              state,
+              established.h,
+              established.encrypt_st
+            )
+
+          {:ok,
+           %Channel{state | channel_state: %Established{established | encrypt_st: encrypt_st}}}
+
+        :responder ->
+          {:ok, %Channel{state | channel_state: established}}
+      end
     end
   end
 
-  # sets vault based on - vault option
-  defp setup_vault(options, data) do
-    case Keyword.get(options, :vault) do
-      nil ->
-        {:error, {:option_is_nil, :vault}}
+  defp handle_decrypted_message(
+         %{onward_route: [], payload: payload} = msg,
+         %Channel{channel_state: %Established{}} = state
+       ) do
+    case ServiceMessage.decode_strict(payload) do
+      {:ok, %ServiceMessage{command: :disconnect}} ->
+        {:stop, :normal, state}
 
-      vault ->
-        {:ok, Map.put(data, :vault, vault)}
+      _error ->
+        {:error, {:unknown_service_msg, msg}}
     end
   end
 
-  # sets peer based on - route option
-  def setup_peer(options, data) do
-    route = Keyword.get(options, :route, [])
-    {:ok, Map.put(data, :peer, %{route: route})}
+  defp handle_decrypted_message(message, %Channel{channel_state: %Established{} = e} = state) do
+    message
+    |> attach_metadata(state.additional_metadata, e)
+    |> Message.trace(state.address)
+    |> Message.forward()
+    |> Router.route()
+
+    {:ok, state}
   end
 
-  # sets initiating_message
-  defp setup_initiating_message(options, data) do
-    case Keyword.get(options, :initiating_message) do
-      nil -> {:ok, data}
-      initiating_message -> {:ok, Map.put(data, :initiating_message, initiating_message)}
-    end
+  defp attach_metadata(msg, additional, %Established{peer_identity: i, peer_identity_id: id}) do
+    Message.with_local_metadata(msg, Map.merge(additional, %{identity: i, identity_id: id}))
   end
 
-  defp setup_callback_route(options, data) do
-    case Keyword.get(options, :callback_route) do
-      nil ->
-        {:ok, data}
+  defp handle_outer_message_impl(message, %Channel{channel_state: %Established{} = e} = state) do
```

Comment:
Does that crash if it's not in the `Established` state?

Reply:
yes.   I'll change it to discard with a loudly error,   if it do receive a message before handshake is complete is very likely an error.   (at that point,   neither   `create_channel`  nor `start_link_channel`  would had returned back to the caller,  nor responder would have contacted his address anywhere).